app/controllers/training_modules_controller.rb

```diff
@@ -11,4 +11,20 @@ def index
   def show
     @training_module = TrainingModule.find_by(slug: params[:module_id])
   end
-end
+ 
+  #new code
+  def find
+    @find_training_module ||= if is_number?(params[:module_id])
+      TrainingModule.find_by(id: params[:module_id])
+    else
+      TrainingModule.find_by(slug: params[:module_id])
+    end
+    redirect_to "/training/students/#{@find_training_module.slug}"
```

Comment:
`students` here is the path for a specific `TrainingLibrary` with slug `students`. Although it will still render no matter which  value is given for the library slug, it won't accurately reflect the context for the given module.

I think it's important to find the corresponding library as well and reflect that in the URL.

The same concept is even more important for `TrainingSlide`, because a slide URL won't actually render unless it includes the corresponding `TrainingModule` slug as well. (This is because the last layer of routing for training slides is handled on the frontend, and the slide content for the entire module gets loaded in order to display a single slide, so that navigating between slides is instantaneous because the whole module is already loaded.)

Reply:
Do you have to be logged in as a user for the trainingSlide redirect feature to work?

Reply:
No, it shouldn't require being logged in.

Reply:
I have added the corresponding library.slug as well as module slug for the find_training_slide
it works on my end...
would something like this work?

Reply:
I don't think so. It needs to be able to find the right redirect using only one id parameter, so to find the library that a module is in, you will likely need to load all the libraries and find the first one that includes the module's slug as a value within its `categories` data. You can see what this data structure looks like in the files in `training_content/wiki_ed/libraries`.

I think one way to go about this is to add an instead method to the `TrainingModule` file to find and return the corresponding library. (And the same approach should work for `TrainingSlide` to find the corresponding module.)

Reply:
def find_corresponding_library
    @find_training_library ||= if is_number?(params[:module_id])
 TrainingLibrary.all.where("categories.modules.id LIKE ?", params[:module_id])
    else
      TrainingLibrary.all.where("categories.modules.slug LIKE ?", params[:module_id]) 
    end
  end

trying to fetch module slug from the trainingLibrary...but i keep getting slug not found
This error...
[2022-10-28 11:34:39.655 FATAL] NoMethodError (undefined method `slug' for #<TrainingLibrary::ActiveRecord_Relation:0x00007fe39d9dd590>):

Is modules an array inside category?and is category an array column?

Reply:
This should ideally work?

Reply:
Both of those ActiveRecord queries use `where`, so they are returning a Relation (which works similarly to an Array), but you want to find an individual record. I think you can use `find_by` instead to have it return a single record.

Reply:
Okay. I actually worked on it a few days ago, can you review it kindly?